### PR TITLE
chore(telemetry): send event startup errors

### PR DESCRIPTION
**Pull request by Kyle-Verhoog**
Telemetry v2 supports an `error` field in the payload of the `app-started` event which is used to report application start up errors. For this we added the `globalexcept_hook` back in, so that we can report startup exceptions. Support for reporting patching errors on integrations is added as well.

I tested this manually and confirmed that start-up error events do get sent. However it's not currently possible to verify the error's content on metabase or the logs, due to how the event is logged, and how metabase tables have not been migrated to store the error field yet.

Depends on #5501

## Checklist

- [x] Change(s) are motivated and described in the PR description.
- [x] Testing strategy is described if automated tests are not included in the PR.
- [x] Risk is outlined (performance impact, potential for breakage, maintainability, etc).
- [x] Change is maintainable (easy to change, telemetry, documentation).
- [x] [Library release note guidelines](https://ddtrace.readthedocs.io/en/stable/contributing.html#Release-Note-Guidelines) are followed.
- [x] Documentation is included (in-code, generated user docs, [public corp docs](https://github.com/DataDog/documentation/)).
- [x] PR description includes explicit acknowledgement/acceptance of the performance implications of this PR as reported in the benchmarks PR comment.

## Reviewer Checklist

- [x] Title is accurate.
- [x] No unnecessary changes are introduced.
- [x] Description motivates each change.
- [x] Avoids breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes unless absolutely necessary.
- [x] Testing strategy adequately addresses listed risk(s).
- [x] Change is maintainable (easy to change, telemetry, documentation).
- [x] Release note makes sense to a user of the library.
- [x] Reviewer has explicitly acknowledged and discussed the performance implications of this PR as reported in the benchmarks PR comment.

### CODE REVIEW

File: ddtrace/internal/telemetry/writer.py [hunk 1]
```diff
@@ -381,7 +394,13 @@ def _app_started_event(self):
             "dependencies": get_dependencies(),
             "integrations": self._flush_integrations_queue(),
             "configurations": [],
+            "error": {
```
Comment (brettlangdon) [inline ddtrace/internal/telemetry/writer.py:397]: what happens if we get the error after the app-started event?

e.g. we likely will submit the app-started payload before we try to send traces to the agent for the first time

File: ddtrace/internal/telemetry/trace.py [hunk 2]
```diff
@@ -0,0 +1,10 @@
+"""
+Trace-specific telemetry.
+"""
+
+
+# Error codes to identify specific types of errors
+# To support multiple errors occurring, each error
```
Comment (ZStriker19) [inline ddtrace/internal/telemetry/trace.py:7]:
```suggestion
# Each error
```
in `writer.py` we say that only a single error can be reported.

File: ddtrace/internal/telemetry/writer.py (hunk 1 above)
Comment (ZStriker19) [reply ddtrace/internal/telemetry/writer.py:397]: Will the telemetry writer not write this then? https://github.dev/DataDog/dd-trace-py/pull/5500/files#diff-0cb9315daf5ed5208de657467cc10a6b7dda616a161c9b5b1f8cff83c95894deR129

**Review by ZStriker19 — CHANGES_REQUESTED** [review]

File: tests/telemetry/test_telemetry.py [hunk 3]
```diff
@@ -154,3 +155,32 @@ def test_logs_after_fork(run_python_code_in_subprocess):
 
     assert status == 0, err
     assert err == b""
+
+
```
Comment (ZStriker19) [inline tests/telemetry/test_telemetry.py:159]: should we add a test for telemetry event generated when the tracer fails to write to the agent?

File: ddtrace/internal/telemetry/writer.py (hunk 1 above)
Comment (brettlangdon) [reply ddtrace/internal/telemetry/writer.py:397]: We only ever call `_app_started_event()` from `_start_service`. And we start the telemetry writer when we start the agent writer.

It seems like we delay start the telemetry writer? We only enable it when the trace writer is enabled?

https://github.com/DataDog/dd-trace-py/blob/a21545212/ddtrace/internal/writer.py#L695-L699

So my concerns might not be an issue.

I do hate the explicit startup relationship here, but... should be ok for now how it is written.

File: ddtrace/internal/writer.py [hunk 4]
```diff
@@ -458,16 +459,17 @@ def flush_queue(self, raise_exc=False):
                 self._metrics_dist("http.errors", tags=["type:err"])
                 self._metrics_dist("http.dropped.bytes", len(encoded))
                 self._metrics_dist("http.dropped.traces", n_traces)
+                log_msg = "failed to send, dropping %d traces to intake at %s after %d retries (%s)" % (
```
Comment (Kyle-Verhoog) [inline ddtrace/internal/writer.py:462]: @ZStriker19 this can be reverted

File: ddtrace/internal/writer.py [hunk 5]
```diff
@@ -458,16 +459,17 @@ def flush_queue(self, raise_exc=False):
...
+                    n_traces,
+                    self._intake_endpoint,
+                    e.last_attempt.attempt_number,
+                    e.last_attempt.exception(),
+                )
+                telemetry_writer.add_error(trace_telemetry.AGENT_ERROR_CODE, log_msg)
                 if raise_exc:
                     e.reraise()
                 else:
-                    log.error(
```
Comment (Kyle-Verhoog) [inline ddtrace/internal/writer.py:464]: patch the DDLogger (internal.logger.DDLogger) instead to add telemetry writer errors

```
    def error(self, *args, **kwargs):
        telemetry_writer.add_error()
        return super(DDLogger, self).error(*args, **kwargs)
```

File: tests/telemetry/test_telemetry.py [hunk 6]
```diff
@@ -154,3 +155,32 @@ def test_logs_after_fork(run_python_code_in_subprocess):
...
+    events = test_agent_session.get_events()
+    assert len(events) == 2
+
+    # Same runtime id is used
+    assert events[0]["runtime_id"] == events[1]["runtime_id"]
+    assert events[0]["request_type"] == "app-closing"
+    assert events[1]["request_type"] == "app-started"
+    assert events[1]["payload"]["error"]["code"] == 2
+    assert events[1]["payload"]["error"]["message"] == "module 'sqlite3' has no attribute 'connect'"
```
Comment (Kyle-Verhoog) [inline tests/telemetry/test_telemetry.py:186]: these will have to be updated to be on the integration object in the payload

File: ddtrace/_monkey.py [hunk 7]
```diff
@@ -150,7 +151,8 @@ def on_import(hook):
         path = "%s.%s" % (prefix, module)
         try:
             imported_module = importlib.import_module(path)
-        except ImportError:
+        except Exception as e:
+            telemetry_writer.add_error(trace_telemetry.INTEGRATION_INSTALL_ERROR_CODE, str(e))
```
Comment (Kyle-Verhoog) [inline ddtrace/_monkey.py:155]: this will be replaced by adding the error to `add_integration` below

File: ddtrace/internal/telemetry/writer.py (hunk 1 above)
Comment (ZStriker19) [reply ddtrace/internal/telemetry/writer.py:397]: Ah Kyle explained. Understood.

File: tests/telemetry/test_telemetry.py (hunk 3 above)
Comment (ZStriker19) [reply tests/telemetry/test_telemetry.py:159]: no because right now this actually does nothing, because we only right the event if it happens on app start and this will never happen on app start.

File: ddtrace/_monkey.py (hunk 7 above)
Comment (ZStriker19) [reply ddtrace/_monkey.py:155]: Can't we just rely on `log.error()`to add the error to the telemetry writer?

File: tests/tracer/test_tracer.py [hunk 8]
```diff
@@ -1934,3 +1935,36 @@ def test_ctx_api():
...
+    e = Foobar()
+
+    tracer = ddtrace.Tracer()
+    tracer._writer.dogstatsd = mock.Mock()
+    with override_global_tracer(tracer):
+        sys.excepthook(e.__class__, e, None)
+    tracer._writer.dogstatsd.increment.assert_has_calls(
+        (mock.call("datadog.tracer.uncaught_exceptions", 1, tags=["class:Foobar"]),)
+    )
+    assert called
```
Comment (brettlangdon) [inline tests/tracer/test_tracer.py:1970]:
```suggestion
    assert called["yes"]
```

?

Comment (ZStriker19) [reply tests/tracer/test_tracer.py]: Actually, I removed the dogstatsd incrementing, from the excepthook. Now instead it creates the app_started event. 

`if asbool(os.getenv("DD_INSTRUMENTATION_TELEMETRY_ENABLED", True)):
            telemetry_writer.add_error(1, tp)
            telemetry_writer.enable()`
            
 Therefore I think that the testing done in test_telemetry.py should cover this.

File: ddtrace/tracer.py [hunk 9]
```diff
@@ -312,6 +313,12 @@ def deregister_on_start_span(self, func):
     def debug_logging(self):
         return log.isEnabledFor(logging.DEBUG)
 
+    def global_excepthook(self, tp, value, traceback):
+        """The global tracer except hook."""
+        if asbool(os.getenv("DD_INSTRUMENTATION_TELEMETRY_ENABLED", True)):
```
Comment (mabdinur) [inline ddtrace/tracer.py:318]: Can we use the config object to store a reference to this environment variable. Something like this: https://github.com/DataDog/dd-trace-py/blob/v1.12.1/ddtrace/settings/config.py#L228

File: ddtrace/internal/telemetry/writer.py [hunk 10]
```diff
@@ -86,12 +87,15 @@ def send_event(self, request):
         except Exception:
             log.debug("failed to send telemetry to the Datadog Agent at %s.", self.url, exc_info=True)
         finally:
-            conn.close()
+            try:
+                conn.close()
+            except UnboundLocalError:
```
Comment (mabdinur) [inline ddtrace/internal/telemetry/writer.py:92]: Why are we handling this exception? This seems like a bug

File: ddtrace/internal/telemetry/writer.py [hunk 11]
```diff
@@ -154,6 +162,13 @@ def add_event(self, payload, payload_type):
             }
             self._events_queue.append(event)
 
+    def add_error(self, code, msg):
+        # type: (int, str) -> None
+        """Add an error to be submitted with an event.
+        Note that this overwrites any previously set errors.
```
Comment (mabdinur) [inline ddtrace/internal/telemetry/writer.py:168]: Is overwriting existing errors the behavior that was defined in the instrumentation telemetry v2 spec? Can we just concatenate the error messages?

File: ddtrace/internal/telemetry/writer.py [hunk 12]
```diff
@@ -124,6 +128,10 @@ def __init__(self, interval):
         self._forked = False  # type: bool
         self._events_queue = []  # type: List[Dict]
         self._lock = forksafe.Lock()  # type: forksafe.ResetObject
+        # Currently telemetry only supports reporting a single error.
+        # If we'd like to report multiple errors in the future
+        # we could hack it in by xor-ing error codes and concatenating strings
+        self._error = (0, "")  # type: Tuple[int, str]
```
Comment (mabdinur) [inline ddtrace/internal/telemetry/writer.py:134]: What are examples of different error codes? What do the different values mean? 
Can we link to the instrumentation telemetry spec here: https://github.com/DataDog/instrumentation-telemetry-api-docs/blob/main/GeneratedDocumentation/ApiDocs/v2/SchemaDocumentation/Schemas/error.md

File: ddtrace/version.py [hunk 13]
```diff
@@ -4,7 +4,7 @@ def get_version():
         from ._version import version
 
         return version
-    except ImportError:
+    except (ImportError, SystemError):
```
Comment (mabdinur) [inline ddtrace/version.py:7]: Why are we catching system errors? Can we add a comment

File: ddtrace/tracer.py [hunk 14]
```diff
@@ -312,6 +313,12 @@ def deregister_on_start_span(self, func):
...
+            telemetry_writer.add_error(1, tp)
+            telemetry_writer.enable()
```
Comment (mabdinur) [inline ddtrace/tracer.py:320]: We now attempt to enable the telemetry_writer in few places (ddtrace/internal/writer.py, ddtrace/__init__.py). Can we do this one place? Ideally in ddtrace/__init__.py

File: ddtrace/__init__.py [hunk 15]
```diff
@@ -1,4 +1,38 @@
...
+
+
+def install_excepthook():
+    """Install a hook that intercepts unhandled exception and send metrics about them."""
+    sys.excepthook = _excepthook
+
+
+def uninstall_excepthook():
+    """Uninstall the global tracer except hook."""
+    sys.excepthook = _ORIGINAL_EXCEPTHOOK
```
Comment (mabdinur) [inline ddtrace/__init__.py:27]: nit: Can we define these helps in some utils/helper module. Would this lead to circular imports

File: ddtrace/internal/telemetry/writer.py (hunk 10 above)
Comment (ZStriker19) [reply ddtrace/internal/telemetry/writer.py:92]: It's because in the code above, we can end up not initializing a `conn` object, before we error out. If that's the case, then I think we'd want to catch the exception rather than crash.

File: ddtrace/internal/telemetry/writer.py (hunk 11 above)
Comment (ZStriker19) [reply ddtrace/internal/telemetry/writer.py:168]: I talked with @Kyle-Verhoog  about the worry of overwriting errors and it was decided that this usually won't happen, and that if it does it's currently acceptable within the telemetry spec we are following. We're not really implement telemetry v2 here, which calls for quite a few additional broader changes.

Comment (ZStriker19) [reply ddtrace/internal/telemetry/writer.py:168]: Telemetry v2 work is scheduled for later.

File: ddtrace/internal/telemetry/writer.py (hunk 12 above)
Comment (ZStriker19) [reply ddtrace/internal/telemetry/writer.py:134]: We're currently only going with two error codes, 1 means error, 0 means no error. We can implement other error codes in the future for telemetry v2 work.

File: ddtrace/tracer.py (hunk 14 above)
Comment (ZStriker19) [reply ddtrace/tracer.py:320]: We need to enable it here, because otherwise it's not started early enough to write the error before the app closes out.

File: ddtrace/__init__.py (hunk 15 above)
Comment (ZStriker19) [reply ddtrace/__init__.py]: Before it was in the tracer module, which did lead to circular imports, not sure if utils/helper module would have the same result.

File: ddtrace/tracer.py (hunk 14 above)
Comment (mabdinur) [reply ddtrace/tracer.py:320]: I see. Previously we queued app started events when the writer was enabled. Now we have to wait until we collect all possible start up errors before sending this event. This can be tricky.

I am not sure if there is an easy way to determine if an application has "successfully started".

Suggestion: We submit an app-started event after the first trace is sent to the agent. Any errors queued after the first trace is submitted will not be considered a start up error and will be ignored by the telemetry writer. The con of this approach is that we need to delay sending all telemetry events until a trace is sent. We can not send metrics/logs/integrations/dependencies/etc until the `app-started` event is ready. This approach could result in a memory leak. A workaround could be to set a max size on the telemetry queue but then we will drop metrics/logs/....

Thoughts? @Kyle-Verhoog

File: ddtrace/internal/telemetry/writer.py (hunk 10 above)
Comment (mabdinur) [reply ddtrace/internal/telemetry/writer.py:92]: Yes if the agent url is invalid this will raise a ValueError and `conn` will be uninitialized. I guess we can initialize `conn to None` before the try-except block. Then so something like this:

```
if conn:
    conn.close()
```

 I think this is cleaner than catching `UnboundLocalError`.

File: ddtrace/version.py (hunk 13 above)
Comment (mabdinur) [inline ddtrace/version.py:7]:
```suggestion
    except ImportError:
```

File: tests/tracer/test_tracer.py [hunk 16]
```diff
@@ -1934,3 +1935,12 @@ def test_ctx_api():
...
         _context.get_items(["appsec.key"])
+
+
+def test_installed_excepthook():
+    telemetry.install_excepthook()
+    assert sys.excepthook is telemetry._excepthook
+    telemetry.uninstall_excepthook()
+    assert sys.excepthook is not telemetry._excepthook
+    telemetry.install_excepthook()
+    assert sys.excepthook is telemetry._excepthook
```
Comment (mabdinur) [inline tests/tracer/test_tracer.py:1946]:
```suggestion
    assert sys.excepthook is telemetry._excepthook
    # Reset exception hooks
    telemetry.uninstall_excepthook()
```

File: tests/telemetry/test_telemetry.py [hunk 17]
```diff
@@ -154,3 +156,78 @@ def test_logs_after_fork(run_python_code_in_subprocess):
...
+patch(raise_errors=False, sqlite3=True)
+tracer.trace("test").finish()
+tracer.flush()
+"""
+
+    stdout, stderr, status, _ = run_python_code_in_subprocess(code)
+
+    assert status == 0, stderr
+    if sys.version_info[0] < 3:
+        # not quite sure why python 2.7 has a different error here, probably a difference with how it loads modules
```
Comment (mabdinur) [inline tests/telemetry/test_telemetry.py:215]: This is because we're attempting to log errors without calling `logging.basicConfig()`. I'll fix this

File: ddtrace/tracer.py (hunk 14 above)
Comment (mabdinur) [reply ddtrace/tracer.py]: I discussed this issue offline with Mehul. 

1. We will continue to send `app-started` after the first trace is sent to the agent. This will allow us to collect errors associated with connection to the agent. 
2. We can send telemetry logs/metrics as they are queued. There is no need to block until app-started is sent. This will resolve the memory leak issue. Metrics/Log events follow a special pathway and are never mapped to "lifecycle" events (ex: app started/app closed/...).

**Review by mabdinur — DISMISSED** [review]

File: ddtrace/internal/logger.py [hunk 18]
```diff
@@ -125,6 +125,13 @@ def handle(self, record):
         :param record: The log record being logged
         :type record: ``logging.LogRecord``
         """
+        if record.levelno >= logging.ERROR:
+            # avoid circular import
+            from ddtrace.internal.telemetry import telemetry_writer
```
Comment (brettlangdon) [inline ddtrace/internal/logger.py:130]: why doesn't telemetry writer just add a log handler on `ddtrace` logger?

File: ddtrace/internal/telemetry/data.py [hunk 19]
```diff
@@ -76,7 +76,7 @@ def _get_application(key):
         "env": env or "",
         "language_name": "python",
         "language_version": _format_version_info(sys.version_info),
-        "tracer_version": get_version(),
+        "tracer_version": ddtrace.__version__,
```
Comment (brettlangdon) [inline ddtrace/internal/telemetry/data.py:79]: why this change?

File: ddtrace/internal/telemetry/writer.py [hunk 20]
```diff
@@ -86,12 +88,13 @@ def send_event(self, request):
         except Exception:
             log.debug("failed to send telemetry to the Datadog Agent at %s.", self.url, exc_info=True)
         finally:
-            conn.close()
+            if conn:
+                conn.close()
```
Comment (brettlangdon) [inline ddtrace/internal/telemetry/writer.py:92]: is this a fix for a bug in telemetry writer? it seems like it is unrelated to this PR? If so, can you open as it's own PR?

File: ddtrace/internal/logger.py [hunk 21]
```diff
@@ -125,6 +125,13 @@ def handle(self, record):
...
+
+            # currently we only have one error code
+            telemetry_writer.add_error(1, record.msg % record.args)
```
Comment (brettlangdon) [inline ddtrace/internal/logger.py:133]: can we create an enum somewhere with the error codes, I can see how it'll get really easy to accidentally use the same one in two places


also, what would be the meaning of the error code here?

Comment (brettlangdon) [reply ddtrace/internal/logger.py:133]: we also need other metadata here for the error, do we get just the message or do we also get the module/logger that was used? (e.g. was this from `ddtrace._monkey` or `ddtrace.internal.writer` ?

(we'd probably be able to tell from the message, but using the error code or including module would make it much easier to query)

File: ddtrace/internal/telemetry/writer.py [hunk 22]
```diff
@@ -382,10 +389,19 @@ def add_integration(self, integration_name, auto_enabled):
             "enabled": True,
             "auto_enabled": auto_enabled,
             "compatible": True,
-            "error": "",
+            "error": self._error[1],  # the integration error only takes a message, no code
```
Comment (brettlangdon) [inline ddtrace/internal/telemetry/writer.py:392]: is this any error or the error specifically for this integration?

what if the last captured error is from the writer? it wouldn't make sense here

File: ddtrace/internal/telemetry/writer.py [hunk 23]
```diff
@@ -382,10 +389,19 @@ def add_integration(self, integration_name, auto_enabled):
...
         }
+        # Reset the error after it has been reported.
+        self._error = (0, "")
         self._integrations_queue.append(integration)
 
+    def add_error(self, code, msg):
```
Comment (brettlangdon) [inline ddtrace/internal/telemetry/writer.py:398]: what happens if we set an error after the app started event is sent?

Comment (mabdinur) [reply ddtrace/internal/telemetry/writer.py:398]: It gets stored and never sent. We will need to implement the `app-configurations-changed` event to send errors after startup. This is out of scope for now.

File: ddtrace/internal/logger.py (hunk 21 above)
Comment (mabdinur) [reply ddtrace/internal/logger.py:133]: Ahhh that's a good point. I will add the file name and line number to telemetry error messages

File: ddtrace/internal/telemetry/data.py (hunk 19 above)
Comment (mabdinur) [reply ddtrace/internal/telemetry/data.py:79]: To avoid circular imports caused by the exception hooks

File: ddtrace/internal/telemetry/writer.py (hunk 22 above)
Comment (mabdinur) [reply ddtrace/internal/telemetry/writer.py:404]: Yup, I plan to change this we should pass the error directly to add_integrations

File: ddtrace/internal/logger.py (hunk 18 above)
Comment (mabdinur) [reply ddtrace/internal/logger.py:130]: That's a wayyyyy better solution. Thanks for the tip

Comment (mabdinur) [reply ddtrace/internal/logger.py:130]: Logging handlers are mainly used to write logs to different sources. By default handlers acquire an [i/o lock](https://github.com/python/cpython/blob/3.11/Lib/logging/__init__.py#L965) before handling a log. This is expensive and is not necessary to intercept and store log messages for the telemetry writer. Instead of using a stream or file handler I could implement the [nullhandler](https://docs.python.org/3/library/logging.handlers.html#nullhandler) but this is not supported in python 2.7. I could override some of this logic in the file/stream handlers to improve performance but I don't think it's worth the effort.

Another approach could be to define a telemetry log filter. One limitation of this approach is that log filters are not automatically applied to descendant loggers. Ensuring the filter is configured correctly could be a pain.

> Note that filters attached to handlers are consulted before an event is emitted by the handler, whereas filters attached to loggers are consulted whenever an event is logged (using [debug()](https://docs.python.org/3/library/logging.html#logging.debug), [info()](https://docs.python.org/3/library/logging.html#logging.info), etc.), before sending an event to handlers. This means that events which have been generated by descendant loggers will not be filtered by a logger’s filter setting, unless the filter has also been applied to those descendant loggers. ([link](https://docs.python.org/3/library/logging.html#logging.Filter.filter))



Hmmm I can't think of a better alternative. I think this approach mixes concerns but I think it's the easiest to read, maintain, and gives us the best performance. Thoughts? @brettlangdon

File: ddtrace/internal/telemetry/__init__.py [hunk 24]
```diff
@@ -14,3 +16,34 @@
...
+
+def _excepthook(tp, value, traceback):
+    filename = None
+    lineno = None
+    if traceback and traceback.tb_frame and traceback.tb_frame.f_code:
+        lineno = traceback.tb_frame.f_code.co_firstlineno
+        filename = traceback.tb_frame.f_code.co_filename
+    telemetry_lifecycle_writer.add_error(1, str(value), filename, lineno)
+
+    if not telemetry_lifecycle_writer.started and telemetry_lifecycle_writer.enable(start_worker_thread=False):
```
Comment (mabdinur) [inline ddtrace/internal/telemetry/__init__.py:32]:
```suggestion
    if not telemetry_lifecycle_writer.started:
```

File: ddtrace/internal/telemetry/__init__.py [hunk 25]
```diff
@@ -14,3 +16,34 @@
...
+        # Enabling the telemetry writer here causes deadlocks in gunicorn
+        # As a workaround we avoid starting the telemetry worker thread then
+        # manually queue and send events.
```
Comment (mabdinur) [inline ddtrace/internal/telemetry/__init__.py:35]:
```suggestion
        # Starting/stopping the telemetry worker thread in a sys.excepthook causes deadlocks in gunicorn.
        # Here we avoid starting the telemetry worker thread by manually queuing and sending
        # app-started and app-closed events.
        telemetry_lifecycle_writer.enable(start_worker_thread=False)
```

**Review by mabdinur — APPROVED** [review]

File: ddtrace/_monkey.py [hunk 26]
```diff
@@ -157,10 +157,12 @@ def on_import(hook):
         except Exception:
             if raise_errors:
                 raise
-            log.error("failed to import ddtrace module %r when patching on import", path, exc_info=True)
+            error_msg = "failed to import ddtrace module %r when patching on import" % (path,)
+            log.error(error_msg, exc_info=True)
+            telemetry_lifecycle_writer.add_integration(module, False, PATCH_MODULES.get(module) is True, error_msg)
```
Comment (mabdinur) [inline ddtrace/_monkey.py:162]: Previously we only queued integrations that were successfully patched. Now we also collect data about integrations that failed to patch

**Review by mabdinur — DISMISSED** [review]

File: ddtrace/internal/telemetry/__init__.py [hunk 27]
```diff
@@ -14,3 +16,34 @@
 telemetry_lifecycle_writer = TelemetryWriter()
 
 __all__ = ["telemetry_lifecycle_writer"]
+
+
+_ORIGINAL_EXCEPTHOOK = sys.excepthook
+
+
+def _excepthook(tp, value, traceback):
```
Comment (ZStriker19) [inline ddtrace/internal/telemetry/__init__.py:24]: Do we like this in telemetry more than in tracer? Right now it's only used for telemetry, but I can imagine it being used in other places at some point? Probably fine to keep here for now.

File: tests/telemetry/test_telemetry.py [hunk 28]
```diff
@@ -160,3 +161,104 @@ def test_logs_after_fork(run_python_code_in_subprocess):
...
+logging.basicConfig()
+
+import sqlite3
+# patch() of the sqlite integration assumes this attribute is there
+# removing it should cause patching to fail.
+del sqlite3.connect
+
+from ddtrace import patch, tracer
+patch(raise_errors=False, sqlite3=True)
+tracer.trace("test").finish()
```
Comment (ZStriker19) [inline tests/telemetry/test_telemetry.py:242]: Do we need 242 and 243?

Comment (ZStriker19) [reply tests/telemetry/test_telemetry.py:242]: Tested, we do... in order to flush the events?  The writer's `on_shutdown()` doesn't send them?

File: ddtrace/internal/telemetry/__init__.py (hunk 27 above)
Comment (mabdinur) [reply ddtrace/internal/telemetry/__init__.py:24]: I'd prefer to keep this here. Instrumentation Telemetry is not tracer specific. It can be used in the profiler or the debugger.

**Review by ZStriker19 — APPROVED** [review]
